src/lib.rs

```diff
@@ -0,0 +1,564 @@
+            return Err(NodeError::OtherError(anyhow::anyhow!("Finalized block stream dropped.")));
+        }
+    }
+
+    Ok(())
+}
+
+/// Executes service infrastructure. Handles connections to a set of nodes and a
+/// database (configured with `app_config`) on their
+/// own separate threads.
+/// Implementation of `NH` defines hooks used by the node thread, while
+/// implementation of `DH` defines hooks used by the database thread.
+pub async fn run_service<D, P, DH, NH>(
+    sql_schema: &'static str,
+    app_config: SharedIndexerArgs,
+    mut node_hooks: NH,
+    config_logger: Option<impl FnOnce(&mut env_logger::Builder, log::LevelFilter)>,
+) -> Result<(), anyhow::Error>
+where
+    P: PrepareStatements + Send + 'static,
+    D: Send + Sync + 'static,
+    DH: DatabaseHooks<D, P> + Send + Sync + 'static,
+    NH: NodeHooks<D>, {
+    anyhow::ensure!(!app_config.endpoint.is_empty(), "At least one node must be provided.");
+    let db_config = app_config.config;
+
+    let mut log_builder = env_logger::Builder::from_env("TRANSACTION_LOGGER_LOG");
+    log_builder.filter_module(module_path!(), app_config.log_level);
+    if let Some(config_logger) = config_logger {
+        config_logger(&mut log_builder, app_config.log_level);
+    }
+    log_builder.init();
+
+    // This program is set up as follows.
+    // It uses tokio to manage tasks. The reason for this is that it is the
+    // expectation that the main bottleneck is IO, either writing to the database or
+    // waiting for responses from the node.
+    // A background task is spawned whose only purpose is to write to the database.
+    // That task only terminates if the connection to the database is lost and
+    // cannot be established within half a minute or so (duration is governed by
+    // MAX_CONNECT_ATTEMPTS and exponential backoff).
+    //
+    // In the main task we query the nodes for block summaries. If querying the node
+    // fails, or the node is deemed too behind then we try the next node. The given
+    // nodes are tried in sequence and cycled when the end of the sequence is
+    // reached.
+    //
+    // The main task and the database writer task communicate via a bounded channel,
+    // with the main task sending block info and block summaries to the database
+    // writer thread.
+
+    // Since the database connection is managed by the background task we use a
+    // oneshot channel to get the height we should start querying at. First the
+    // background database task is started which then sends the height over this
+    // channel.
+    let (height_sender, height_receiver) = tokio::sync::oneshot::channel();
+    // Create a channel between the task querying the node and the task logging
+    // transactions.
+    let (sender, receiver) = tokio::sync::mpsc::channel(100);
+
+    let stop_flag = Arc::new(AtomicBool::new(false));
+
+    let shutdown_handler_handle = tokio::spawn(set_shutdown(stop_flag.clone()));
```

Comment:
This should be left up to the user and not be in the library. Instead I suggest the `run_service` function to take a `oneshot` (or `watch`) channel for when to shut down